### PR TITLE
@mczernek/notifications suspended after 5 minutes

### PR DESCRIPTION
**Pull request by mczernek**
# Why

Current architecture of expo-notifications is based in many places on invoking `JobIntentService` for each small action. While it provided nice and concise way of handling asynchronous tasks, it cause some of those actions to be deffered until Android decides to perform them. That meant, that after five minutes of inactivity on the phone (blocked and disconnected from charger), those actions were queued for later execution.

One of such actions was handling of just received notification. As a result, after receiving call from remote (FCM) notification was not displayed right away, even though app was being waked by system.

It is described in [#9037](https://github.com/expo/expo/issues/9037)

# How

issue is fixed by converting most of those actions from execution in `JobIntentService` into straight synchronous code.

Mentioned service served not only as as asynchronity provider, but as a way of scoping for Notifications. Since some code in the module is executed without ModuleRegistry, we could not use standard scoping mechanism. As a result, I opted to creating `bare` and `expoview` flavors. Thanks to small modification in `gradle.groovy` this should come unnoticed to users and expo developers. At the same time it provides great way of scoping when we can't use ModuleRegistry.

# Test Plan

`test-suite` and `ncl` - both bare and expo client.

### CODE REVIEW

File: packages/expo-notifications/CHANGELOG.md
Comment (barthap) [inline packages/expo-notifications/CHANGELOG.md:13]: Your Github profile link is incorrect ;)

**Review by esamelson — COMMENTED** [review]
Thanks for jumping on this! My main comment is that it seems a little bit overkill to add the whole expoview flavor and infrastructure to this module just for the one Reconstructor class. Is it possible to just keep all the expoview code in the root `android/src/expoview` directory and override this class from there? I think that as much as possible, the packages should not need to know about the Expo client and when we do need to have special case code for the client, that code should live in the client's codebase rather than in the individual packages. The packages themselves should not need to worry about supporting the Expo client special case and I think it's more confusing to have that code be distributed throughout all the packages rather than collated in one location.

Otherwise it lgtm, but would also want to make sure @lukmccall OKs the architecture changes in this PR before landing, since @sjchmiela is currently OOO.

**Comment by mczernek** [comment]
> Thanks for jumping on this! My main comment is that it seems a little bit overkill to add the whole expoview flavor and infrastructure to this module just for the one Reconstructor class. Is it possible to just keep all the expoview code in the root `android/src/expoview` directory and override this class from there? I think that as much as possible, the packages should not need to know about the Expo client and when we do need to have special case code for the client, that code should live in the client's codebase rather than in the individual packages. The packages themselves should not need to worry about supporting the Expo client special case and I think it's more confusing to have that code be distributed throughout all the packages rather than collated in one location.
> 
> Otherwise it lgtm, but would also want to make sure @lukmccall OKs the architecture changes in this PR before landing, since @sjchmiela is currently OOO.

I do not know of any reliable way of substituting class implementation other than android flavors. We do have mechanism for that based on ModuleRegistry, but it is not applicable, since we do not have access to it.

Another solution I implemented in expo-task-manager is by declaring actual implementation in AndroidManifest's meta-data. I find flavor solutions much friendlier and was considering switching to it in task manager as well, given we got agreement on that.

Another solution I could potentially think of is avoiding scoping reconstructor whatsoever. I guess it would be doable. But instead of walking around the problem I've wanted to try and discuss a solution that seemed clean to me.

**Comment by esamelson** [comment]
Gotcha, thanks for the explanation. I would strongly lean towards the AndroidManifest solution that you used for expo-task-manager, for these reasons:
- it puts the responsibility on expoview for enabling the extra scoping in the managed workflow (which should be treated as the special case here), rather than putting responsibility on react-native-unimodules to hide the extra flavor in bare workflow projects (which are the general case)
- maintainability for developers touching the expoview code in the future. Confusion about how the class is being overridden in AndroidManifest can be headed off by explaining it in comments, but my opinion is that expoview as a library becomes harder to maintain if it's distributed around multiple directories/packages in this repo.

Happy to discuss/consider this further though!

File: apps/bare-expo/android/app/src/main/java/dev/expo/payments/generated/BasePackageList.java
Comment (lukmccall) [inline apps/bare-expo/android/app/src/main/java/dev/expo/payments/generated/BasePackageList.java:62]: I'm not sure if this line should be in this PR ;)

File: packages/expo-notifications/CHANGELOG.md
Comment (lukmccall) [inline packages/expo-notifications/CHANGELOG.md:13]:
```suggestion
- Fix notifications not being displayed after five minutes of phone inactivity on Android. ([#9287](https://github.com/expo/expo/pull/9287) by [@mczernek](https://github.com/mczernek))
```

🧹

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/service/ExpoNotificationSchedulerService.java
Comment (lukmccall) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/service/ExpoNotificationSchedulerService.java:354]:
```suggestion
   * @param context Context this is being called from
```
🧹

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/interfaces/NotificationsReconstructor.java
Comment (lukmccall) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/interfaces/NotificationsReconstructor.java:11]: Why did you add the `getName()` function?

**Review by lukmccall — COMMENTED** [review]
Look good :D 

I've left some comments ;)

**Comment by mczernek** [comment]
> Gotcha, thanks for the explanation. I would strongly lean towards the AndroidManifest solution that you used for expo-task-manager, for these reasons:
> 
> * it puts the responsibility on expoview for enabling the extra scoping in the managed workflow (which should be treated as the special case here), rather than putting responsibility on react-native-unimodules to hide the extra flavor in bare workflow projects (which are the general case)
> * maintainability for developers touching the expoview code in the future. Confusion about how the class is being overridden in AndroidManifest can be headed off by explaining it in comments, but my opinion is that expoview as a library becomes harder to maintain if it's distributed around multiple directories/packages in this repo.
> 
> Happy to discuss/consider this further though!

I see your point and I could go with that. Will just have another try with convincing you to my solution :)

Flavours are very widely used concept in Android world, so it should not come with any surprise to those coming from Android. I believe we can get used to it as well :)
The only responsibility of react-native-unimodules is to set strategy for when flavour dimension is missing. In my humble opinion, that's much clearer than having class name typed in AndroidManifest's meta-data and recreating it with reflection.

Moreover, I am hoping that one day all scoping might be done with flavours. That would give us uniform and clear scoping with no need for reflection for that. Moreover, I can imagine in further future, that we just switch from client to standalone app by selecting the right flavour in expo repository.

Last but not least, I am all for having all module related code within the module. I believe it is much easier, while developing, say, `expo-notifications` to stick with `expo-notifications` module, rather than switching from respective module to expoview back and forth. I can see that is a matter of preference, though.

The way I see it, none of the solutions has clear advantage over the other. While I prefer mine, I can totally agree on rewriting it to yours given I failed to convince you :)

**Comment by esamelson** [comment]
Your points make sense as well 😄 but I still want to push back a little:

> Flavours are very widely used concept in Android world, so it should not come with any surprise to those coming from Android. I believe we can get used to it as well :)
> The only responsibility of react-native-unimodules is to set strategy for when flavour dimension is missing. In my humble opinion, that's much clearer than having class name typed in AndroidManifest's meta-data and recreating it with reflection.

My hesitation is not to the flavors themselves but to the fact that we would be shipping internal-only flavors in an external-facing package. We've seen that when complications due to our client codebase sneak into externally-consumed packages, this tends to make things weirder and more difficult for developers (ExpoKit is an extreme example, but even the unimodule architecture itself is something we want to reduce), so I would strongly prefer to keep the packages as lean and close to other plain RN libraries as possible. However, if this type of practice (shipping flavors used for internal builds) is common among Android libraries I could be convinced otherwise! 😄 

> Last but not least, I am all for having all module related code within the module. I believe it is much easier, while developing, say, `expo-notifications` to stick with `expo-notifications` module, rather than switching from respective module to expoview back and forth. I can see that is a matter of preference, though.

This makes a lot of sense. One point in favor of keeping the scoped expoview code together, though, is that this is the convention we already follow for all the other scoped modules (which use ModuleRegistry), so this one would be a bit of an outlier.

**Comment by mczernek** [comment]
As of your concern with shipping flavour specific code to users, we try to mitigate that by not publishing expoview specific code to npm in modules (as seen in all `.npmignore` files). There is still a risk that users might face problems here, especially when they try to set up unimodules on their own (without unimodules gradle plugin).

To sum up this discussion, we do not have one approach that would solve all the issues, but you still prefer yours to mine :) I am comfortable to migrate to yours then, just let me know my assumption is correct.

**Comment by esamelson** [comment]
OK, yeah, that sounds good. Thanks for your willingness to discuss/change this!

**Review by cruzach — COMMENTED** [review]
I'm not as familiar as eric and lukasz with the rest of the `expoview` and `expo-notifications` infrastructure, respectively, so I defer to them for that, but this doesn't look like it breaks any of the recent notification category functionality, thanks 👍

File: android/app/build.gradle
Comment (lukmccall) [inline android/app/build.gradle:186]:
```suggestion
```
🧹

File: android/expoview/src/main/AndroidManifest.xml
Comment (lukmccall) [inline android/expoview/src/main/AndroidManifest.xml:97]: I'm not sure if this change should be in this PR ;)

File: packages/expo-notifications/android/build.gradle
Comment (lukmccall) [inline packages/expo-notifications/android/build.gradle:42]:
```suggestion
```
❔

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/NotificationsScoper.java
Comment (lukmccall) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/NotificationsScoper.java:17]: I'm not a fan of this syntactic sugar. Hovewer, it can stay ;)

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/interfaces/NotificationsBuilderCreator.java
Comment (lukmccall) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/interfaces/NotificationsBuilderCreator.java:8]:
```suggestion
@FunctionalInterface
public interface NotificationsBuilderCreator {
```

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/interfaces/NotificationsReconstructor.java
Comment (lukmccall) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/interfaces/NotificationsReconstructor.java:7]:
```suggestion
@FunctionalInterface
public interface NotificationsReconstructor {
```

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/service/ExpoNotificationSchedulerService.java
Comment (lukmccall) [reply packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/service/ExpoNotificationSchedulerService.java:354]: ping

File: packages/react-native-unimodules/gradle.groovy
Comment (lukmccall) [inline packages/react-native-unimodules/gradle.groovy:224]:
```suggestion

```

**Review by lukmccall — APPROVED** [review]
LGTM 🚀🚀🚀

File: android/expoview/src/main/AndroidManifest.xml
Comment (mczernek) [reply android/expoview/src/main/AndroidManifest.xml:97]: Wow, that's weird. Must be some side effect of rebasing/upmerging. Will take a look at that.

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/interfaces/NotificationsReconstructor.java
Comment (mczernek) [reply packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/interfaces/NotificationsReconstructor.java:7]: This one, however, is not intended to be Functional. It just so happens it has one method. But I guess we should use this annotation only when we really want it to stay this way, while I have nothing against expanding this interface in the future, according to our needs.

File: packages/expo-notifications/android/build.gradle
Comment (mczernek) [reply packages/expo-notifications/android/build.gradle:42]: Leftover, thanks, will delete.

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/interfaces/NotificationsReconstructor.java
Comment (mczernek) [reply packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/interfaces/NotificationsReconstructor.java]: Just for debugging, removed :)

File: android/expoview/src/main/AndroidManifest.xml
Comment (mczernek) [reply android/expoview/src/main/AndroidManifest.xml:97]: Really, there is only change in formatting. But GH shows the diff quite weirdly.

**Pull request merged**
